kernel/memory_structs/src/lib.rs

```diff
@@ -473,45 +607,50 @@ macro_rules! implement_page_frame_range {
-                pub fn overlap(&self, other: &$TypeName) -> Option<$TypeName> {
-                    let starts = max(*self.start(), *other.start());
-                    let ends   = min(*self.end(),   *other.end());
-                    if starts <= ends {
-                        Some($TypeName::new(starts, ends))
-                    } else {
-                        None
-                    }
+                #[doc = "Returns the number of 4KiB chunks in this range. \
+                    Needed for functions in the impl block generic over chunk size P, as the compiler cannot recognize which methods to call."]
+                const fn [<size_in_ $chunk:lower s_gen>](&self) -> usize {
+                   // add 1 because it's an inclusive range
+                   (self.0.end().number + 1).saturating_sub(self.0.start().number)
                 }
 
-                #[doc = "Returns `true` if the `other` `" $TypeName "` is fully contained within this `" $TypeName "`."]
-                pub fn contains_range(&self, other: &$TypeName) -> bool {
-                    !other.is_empty()
-                    && (other.start() >= self.start())
-                    && (other.end() <= self.end())
+                #[doc = "Changes this `" $TypeName "` to have a size of 4KiB. This does not perform any alignment. \
+                It simply changes the marker type for usage with functions that want a range of default-sized pages."]
+                pub fn as_4kb_range(&self) -> $TypeName {
+                    $TypeName(RangeInclusive::new(self.start().as_4kb(), self.end().as_4kb()))
                 }
             }
-            impl fmt::Debug for $TypeName {
+            impl $TypeName<Page2MiB> {
+                #[doc = "Returns the number of [`" $chunk "`]s covered by this iterator.\n\n \
+                Use this instead of [`Iterator::count()`] method. \
+                This is instant, because it doesn't need to iterate over each entry, unlike normal iterators."]
+                pub const fn [<size_in_ $chunk:lower s>](&self) -> usize {
+                   // add 1 because it's an inclusive range; divide by 512 because it's a 2MiB page
+                   (self.0.end().number + 1).saturating_sub(self.0.start().number) / 512
+                }
+            }
+            impl $TypeName<Page1GiB> {
+                #[doc = "Returns the number of [`" $chunk "`]s covered by this iterator.\n\n \
+                Use this instead of [`Iterator::count()`] method. \
+                This is instant, because it doesn't need to iterate over each entry, unlike normal iterators."]
+                pub const fn [<size_in_ $chunk:lower s>](&self) -> usize {
+                   // add 1 because it's an inclusive range; divide by 262144 because it's a 2MiB page
+                   (self.0.end().number + 1).saturating_sub(self.0.start().number) / (512 * 512)
+                }
+            }
```

Comment:
all of these functions can be made generic by using the `P` parameter. You don't need special functions for any specific page size.